Lesson4.ipynb

```diff
@@ -29,6 +32,655 @@
+    "            teamTotalSteps = teamTotalSteps + totalStepsByRow(row, header)\n",
+    "```\n",
+    "** **\n",
+    "### Loop for team names\n",
+    "\n",
+    "To print the total steps for each team, let's do a loop for the team names outside the code block above.\n",
+    "\n",
+    "Please refer the code block below:\n",
+    "\n",
+    "```python\n",
+    "for name in team_names:\n",
+    "    totalStepsByTeam = getTotalStepsByTeamNameFromCSV(name)\n",
+    "```\n",
+    "\n",
+    "** **\n",
+    "\n",
+    "***Tips: When using csv.reader() or csv.DictReader() in Python, all the data only can be read one time. After reading, the data will be deleted in memory. If you want to use the data for multi-times, you need to call the reader() or DictReader() which means read CSV file for each time.***\n",
+    "\n",
+    "### Reference\n",
+    "\n",
+    "** Entire code for printing total steps by each team**\n",
+    "\n",
+    "Please refer to the code block below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 127,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Total steps of Team TBD is: 1820997\n",
+      "Total steps of Team Sole Striders is: 954382\n",
+      "Total steps of Team FuFu is: 1198137\n",
+      "Total steps of Team Tears for Beers is: 1468397\n",
+      "Total steps of Team Stop when you drop!! is: 1721534\n",
+      "Total steps of Team Cereal Killers is: 1333977\n",
+      "Total steps of Team Here come the hotsteppers is: 1107342\n",
+      "Total steps of Team The Slimsons is: 1484755\n",
+      "Total steps of Team Just Step It is: 1128860\n",
+      "Total steps of Team Will Walk for Wine is: 1158732\n",
+      "Total steps of Team 11,111 to No. 1 is: 1395549\n",
+      "Total steps of Team Scrambled Legs is: 1480327\n",
+      "Total steps of Team Smarty Pints is: 1055281\n",
+      "Total steps of Team  is: 939917\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Print total steps by each team\n",
+    "import csv\n",
+    "\n",
+    "def totalStepsByRow(row, header):\n",
+    "    totalSteps = 0\n",
+    "    for header_index, fieldNames in enumerate(header):\n",
+    "        steps_in_column = row[header[header_index]]\n",
+    "        if header_index > 3 and steps_in_column != '':\n",
+    "            totalSteps = totalSteps + int(steps_in_column)\n",
+    "    return totalSteps\n",
+    "\n",
+    "def getTeamNamesFromCSV():\n",
+    "    with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "        readCSV = csv.DictReader(csvfile)\n",
+    "        team_names = []\n",
+    "        for row in readCSV:\n",
+    "            if row['team_name'] not in team_names:\n",
+    "                team_names.append(row['team_name'])\n",
+    "    return team_names\n",
+    "\n",
+    "def getTotalStepsByTeamNameFromCSV(team_name):\n",
+    "    with open('Biggest Loser 2018.csv') as csvfile:\n",
```

Comment:
either double open file or we can use list casting:
```Python
# Print total steps by each team
import csv

def totalStepsByRow(row, header):
    totalSteps = 0
    for header_index, fieldNames in enumerate(header):
        steps_in_column = row[header[header_index]]
        if header_index > 3 and steps_in_column != '':
            totalSteps = totalSteps + int(steps_in_column)
    return totalSteps

def get_team_names(steps_data):
    team_names = []
    for row in steps_data:
        if row['team_name'] not in team_names:
            team_names.append(row['team_name'])
    return team_names

def get_total_steps(team_name, steps_data, header):
    total_steps = 0
    for row in steps_data:
        if row['team_name'] == team_name:
            total_steps = total_steps + totalStepsByRow(row, header)
    return total_steps

with open('Biggest Loser 2018.csv') as csvfile:
    csvdata = csv.DictReader(csvfile)                           # Returns an iterator object, and iterator can only loop once!
    header = csvdata.fieldnames                                 # Get the header
    steps_data = list(csvdata)                                  # Convert the steps data to list since we loop over more than once
    team_names = get_team_names(steps_data)
    for name in team_names:
        team_total_steps = get_total_steps(name, steps_data, header)
        print('Total steps of Team ' + name + ' is: ' + str(team_total_steps))
```

Reply:
Can add an open-ended question as a challenge like: Currently our simple implementation loops through the data multiple times (18 teams, 1 for getting team name and once for each team). This is not very efficient because we keep revisits the same data. Imagine we have more teams the number of steps will keep growing. Can we reduce the number of loops?